Feature: `make check`

reneme opened the pull request:
This adds a build target `make check` in response of [#2300](https://github.com/randombit/botan/issues/2300). Invoking `make check` simply builds the test binary and runs it without further command line parameters. Configuring the dynamic library loading (via `[DY]LD_LIBRARY_PATH`) is taken care of automatically -- after [#2294](https://github.com/randombit/botan/pull/2294) introduced some inconvenience in that regard.

 For now, this is solely a user-facing build target and is not used in CI. The `ci_build.py` script makes use of the `./botan-test` binary's command line parameters for fine-grained control and cannot easily be adapted to use `make check`.

reneme commented on src/scripts/check.py:
```diff
@@ -0,0 +1,89 @@
+    if is_macos():
+        env["DYLD_LIBRARY_PATH"] = os.path.abspath(".")
+    elif is_linux():
+        env["LD_LIBRARY_PATH"] = os.path.abspath(".")
```
Do we need more platforms here?

randombit replied:
Most other OSes (eg FreeBSD) also use LD_LIBRARY_PATH. Instead of trying to detect which to use, just always set both vars. Also need to handle existing value of these environments instead of overwriting them, eg if OpenSSL is in some side dir but user has adjusted with LD_LIBRARY_PATH.

randombit commented on src/build-data/makefile.in:
```diff
@@ -59,6 +59,14 @@ distclean:
+%{endif}
+%{unless build_shared_lib}
+	$(PYTHON_EXE) $(SCRIPTS_DIR)/check.py --test-exe="%{test_exe}"
+%{endif}
```
Instead of this, pass to `check.py` the build directory via `--build-dir` and then load `build_config.json` to get test exe, shared lib usage etc

reneme commented:
Thanks for the pointer to the previously closed pull request. I took the `--test-data` parameter handling from there. Furthermore, I switched to loading `build_config.json`, removed `rpath=$ORIGIN` and added `check.py` to pylint.

reneme commented:
Flaky test on macOS?
```
  ERROR: Test cli_tls_socket_tests threw exception: [Errno 32] Broken pipe
```

randombit commented:
Flaky test. I restarted and it passed.
